### PR TITLE
Market decentralized - create agreement

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves:  #365

Based on PR: https://github.com/golemfactory/yagna/pull/398
Change base to master before merging.

I can't test this PR without counter proposal on provider side.

### CODE REVIEW

File: core/market/decentralized/src/db/models/agreement.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+pub type AgreementId = ProposalId;
+
+/// TODO: Could we avoid having separate enum type for database
+///  and separate for client?
+#[derive(FromPrimitive, AsExpression, FromSqlRow, PartialEq, Debug, Clone, Copy)]
+#[sql_type = "Integer"]
+pub enum AgreementState {
+    /// Newly created by a Requestor (based on Proposal)
+    Proposal = 0,
+    /// Confirmed by a Requestor and send to Provider for approval
```
Comment (jiivan) [inline core/market/decentralized/src/db/models/agreement.rs:25]:
```suggestion
    /// Confirmed by a Requestor and sent to Provider for approval
```

File: core/market/decentralized/src/db/models/agreement.rs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    pub demand_properties: String,
+    pub demand_constraints: String,
+
+    pub provider_id: NodeId,
+    pub requestor_id: NodeId,
+
+    /// End of validity period.
+    /// Agreement needs to be accepted, rejected or cancelled before this date; otherwise will expire.
+    pub creation_ts: NaiveDateTime,
+    pub valid_to: NaiveDateTime,
```
Comment (jiivan) [inline core/market/decentralized/src/db/models/agreement.rs:56]:
```suggestion
    pub valid_till: NaiveDateTime,
```

File: core/market/decentralized/src/db/models/agreement.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    pub demand_properties: String,
+    pub demand_constraints: String,
+
+    pub provider_id: NodeId,
+    pub requestor_id: NodeId,
+
+    /// End of validity period.
+    /// Agreement needs to be accepted, rejected or cancelled before this date; otherwise will expire.
+    pub creation_ts: NaiveDateTime,
```
Comment (jiivan) [inline core/market/decentralized/src/db/models/agreement.rs:55]: `ts` suffix is misleading for me. It's a datetime not a float/int.

Comment (tworec) [reply core/market/decentralized/src/db/models/agreement.rs:55]: Make a new issue, and leave in this PR as is.

Comment (tworec) [reply core/market/decentralized/src/db/models/agreement.rs:55]: To be compatible with rest of yagna modules we decided to use `_date` suffix

Comment (tworec) [reply core/market/decentralized/src/db/models/agreement.rs:55]: #416

Comment (jiivan) [reply core/market/decentralized/src/db/models/agreement.rs:55]: ok

File: core/market/decentralized/src/db/models/agreement.rs (hunk 2 above)
Comment (tworec) [reply core/market/decentralized/src/db/models/agreement.rs:56]: added to #416

File: core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql [hunk 4]
```diff
@@ -95,4 +95,42 @@ CREATE TABLE market_proposal(
...
+    creation_ts DATETIME NOT NULL,
+    valid_to DATETIME NOT NULL,
+	state INTEGER NOT NULL,
+	approved_date DATETIME,
+
+	proposed_signature TEXT,
+	approved_signature TEXT,
+	committed_signature TEXT,
+    FOREIGN KEY(state) REFERENCES agreement_state (id)
+);
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/migrations/2020-05-15-093513_init-db/up.sql:119]: Good point, but no.
Until we have no users of our crate (it is behind compilation flag) we need not to do it as separate migration

**Pull request merged**